sql/sqlite/driver.go

```diff
@@ -181,19 +186,89 @@ func openTx(ctx context.Context, db *sql.DB, opts *sql.TxOptions) (*sqlclient.Tx
 	if err != nil {
 		return nil, err
 	}
+	before, err := violations(ctx, tx)
+	if err != nil {
+		return nil, err
+	}
 	return &sqlclient.Tx{
 		Tx: tx,
-		Close: func() error {
-			if on.Bool {
-				if _, err := db.ExecContext(ctx, "PRAGMA foreign_keys = on"); err != nil {
-					return fmt.Errorf("sql/sqlite: set 'foreign_keys = on': %w", err)
+		Commit: func() error {
+			after, err := violations(ctx, tx)
+			if err != nil {
+				if err2 := tx.Rollback(); err2 != nil {
+					err = fmt.Errorf("%v: %w", err2, err)
+				}
+				return enableFK(ctx, db, on.Bool, err)
+			}
+			if vs := newViolations(before, after); len(vs) > 0 {
+				err := fmt.Errorf("sql/sqlite: foreign key mismatch: %+v", vs)
+				if err2 := tx.Rollback(); err2 != nil {
+					err = fmt.Errorf("%v: %w", err2, err)
 				}
+				return enableFK(ctx, db, on.Bool, err)
 			}
-			return nil
+			return enableFK(ctx, db, on.Bool, tx.Commit())
+		},
+		Rollback: func() error {
+			return enableFK(ctx, db, on.Bool, tx.Rollback())
 		},
 	}, nil
 }
 
+func enableFK(ctx context.Context, db *sql.DB, do bool, err error) error {
+	if do {
+		// Re-enable foreign key checks if they were enabled before.
+		if _, err2 := db.ExecContext(ctx, "PRAGMA foreign_keys = on"); err2 != nil {
+			err2 = fmt.Errorf("sql/sqlite: set 'foreign_keys = on': %w", err2)
+			if err != nil {
+				return fmt.Errorf("%v: %w", err2, err)
+			}
+			return err2
+		}
+	}
+	return err
+}
+
+func violations(ctx context.Context, tx *sql.Tx) ([]violation, error) {
+	rows, err := tx.QueryContext(ctx, "PRAGMA foreign_key_check")
+	if err != nil {
+		return nil, fmt.Errorf("sql/sqlite: querying 'foreign_key_check' pragma: %w", err)
+	}
+	defer rows.Close()
+	var vs []violation
+	for rows.Next() {
+		var v violation
+		if err := rows.Scan(&v.tbl, &v.row, &v.ref, &v.index); err != nil {
+			return nil, fmt.Errorf("sql/sqlite: querying 'foreign_key_check' pragma: scanning rows: %w", err)
+		}
+		vs = append(vs, v)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("sql/sqlite: querying 'foreign_key_check' pragma: scanning rows: %w", err)
+	}
+	return vs, nil
+}
+
+// equalViolations compares the foreign key violations before starting a transaction with the ones afterwards.
+// It returns violations found in v2 that are not in v1.
+func newViolations(v1, v2 []violation) (vs []violation) {
+	for _, v := range v2 {
+		if !contains(v1, v) {
+			vs = append(vs, v)
+		}
```

Comment:
`slices.Index` can be used here, but we don't use `slices` as a dependency in this module (only in `cmd/atlas`).